### PR TITLE
Pil 330 - Enter registered org name and address of NFM

### PR DESCRIPTION
**Pull request by javed-iqbal1**

### CODE REVIEW

File: conf/application.conf [hunk 1]
```diff
@@ -36,7 +36,11 @@ play.modules.enabled += "uk.gov.hmrc.play.bootstrap.AuthModule"
 play.modules.enabled += "uk.gov.hmrc.play.bootstrap.graphite.GraphiteMetricsModule"
 play.modules.enabled += "uk.gov.hmrc.mongo.play.PlayMongoModule"
 play.modules.enabled += "config.Module"
-
+location.canonical.list {
+  all = "location-autocomplete-canonical-list.json"
+  EUAndEEA = "location-canonical-list-EUAndEEA.json"
```
Comment (AKKUM) [inline conf/application.conf:41]: Do we need this EUAndEEA?

File: conf/application.conf [hunk 2]
```diff
@@ -36,7 +36,11 @@ play.modules.enabled += "uk.gov.hmrc.play.bootstrap.AuthModule"
...
+}
+validCountryCodes = "AD,AE,AF,AG,AI,AL,AM,AN,AO,AQ,AR,AS,AT,AU,AW,AX,AZ,BA,BB,BD,BE,BF,BG,BH,BI,BJ,BL,BM,BN,BO,BQ,BR,BS,BT,BV,BW,BY,BZ,CA,CC,CD,CF,CG,CH,CI,CK,CL,CM,CN,CO,CR,CS,CU,CV,CW,CX,CY,CZ,DE,DJ,DK,DM,DO,DZ,EC,EE,EG,EH,ER,ES,ET,EU,FC,FI,FJ,FK,FM,FO,FR,GA,GB,GD,GE,GF,GG,GH,GI,GL,GM,GN,GP,GQ,GR,GS,GT,GU,GW,GY,HK,HM,HN,HR,HT,HU,ID,IE,IL,IM,IN,IO,IQ,IR,IS,IT,JE,JM,JO,JP,KE,KG,KH,KI,KM,KN,KP,KR,KW,KY,KZ,LA,LB,LC,LI,LK,LR,LS,LT,LU,LV,LY,MA,MC,MD,ME,MF,MG,MH,MK,ML,MM,MN,MO,MP,MQ,MR,MS,MT,MU,MV,MW,MX,MY,MZ,NA,NC,NE,NF,NG,NI,NL,NO,NP,NR,NT,NU,NZ,OM,OR,PA,PE,PF,PG,PH,PK,PL,PM,PN,PR,PS,PT,PW,PY,QA,RE,RO,RS,RU,RW,SA,SB,SC,SD,SE,SG,SH,SI,SJ,SK,SL,SM,SN,SO,SR,SS,ST,SV,SX,SY,SZ,TC,TD,TF,TG,TH,TJ,TK,TL,TM,TN,TO,TP,TR,TT,TV,TW,TZ,UA,UG,UM,UN,US,UY,UZ,VA,VC,VE,VG,VI,VN,VU,WF,WS,YE,YT,ZA,ZM,ZW"
```
Comment (AKKUM) [inline conf/application.conf:43]: Can this be populated from File. Otherwise we will have to maintain two places.

File: conf/messages.en [hunk 3]
```diff
@@ -259,3 +259,41 @@ isNFMUKBased.heading.caption = Business details
...
+nfmNameRegistrationController.title = What is the name of the nominated filing member?
+nfmNameRegistrationController.heading.caption = Business details
+nfmNameRegistrationController.heading = What is the name of the nominated filing member?
+nfmNameRegistrationController.checkYourAnswersLabel = What is the name of the nominated filing member?
+nfmNameRegistrationController.error.required =  You need to enter the name of the nominated filing member
+nfmNameRegistrationController.error.length = The name of the nominated filing member must be 105 characters or less
+nfmNameRegistrationController.change.hidden = What is the name of the nominated filing member?
+
+
+nfm-registered-address.title=Where is the registered office address of {0}?
```
Comment (AKKUM) [inline conf/messages.en:272]: Why did we put - in message file?

File: test/controllers/fm/NfmRegisteredAddressControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import play.api.test.FakeRequest
+import play.api.test.Helpers._
+import utils.InputOption
+
+import scala.concurrent.Future
+
+class NfmRegisteredAddressControllerSpec extends SpecBase {
+  val formProvider = new NfmRegisteredAddressFormProvider()
+
+  def controller(): NfmRegisteredAddressController =
```
Comment (AKKUM) [inline test/controllers/fm/NfmRegisteredAddressControllerSpec.scala:37]: Same here do we need this?

File: app/controllers/fm/NfmNameRegistrationController.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                                    isNFMnStatus = RowStatus.InProgress,
+                                    withoutIdRegData = Some(WithoutIdNfmData(registeredFmName = value))
+                                  )
+                                )
+                              )
+            _ <- userAnswersConnectors.save(updatedAnswers.id, Json.toJson(updatedAnswers.data))
+          } yield Redirect(controllers.fm.routes.NfmRegisteredAddressController.onPageLoad(mode))
+      )
+  }
+}
```
Comment (AKKUM) [inline app/controllers/fm/NfmNameRegistrationController.scala:83]: Same bookmark issue here?

File: app/controllers/fm/NfmRegisteredAddressController.scala [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            _ <- userAnswersConnectors.save(updatedAnswers.id, Json.toJson(updatedAnswers.data))
+          } yield Redirect(controllers.routes.UnderConstructionController.onPageLoad)
+      )
+  }
+  private def getUserName(request: DataRequest[AnyContent]): String = {
+    val fmDetails = request.userAnswers.get(NominatedFilingMemberPage)
+    fmDetails.fold("")(fmData => fmData.withoutIdRegData.fold("")(withoutId => withoutId.registeredFmName))
+  }
+
+}
```
Comment (AKKUM) [inline app/controllers/fm/NfmRegisteredAddressController.scala:93]: Same here - Book mark

File: app/forms/NfmNameRegistrationControllerFormProvider.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import javax.inject.Inject
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+class NfmNameRegistrationControllerFormProvider @Inject() extends Mappings {
+
+  def apply(): Form[String] =
+    Form(
+      "value" -> text("nfmNameRegistrationController.error.required")
```
Comment (AKKUM) [inline app/forms/NfmNameRegistrationControllerFormProvider.scala:28]: are you sure we require 1000 char validation for name?

File: app/forms/NfmNameRegistrationControllerFormProvider.scala [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        .verifying(maxLength(1000, "nfmNameRegistrationController.error.length"))
```
Comment (AKKUM) [inline app/forms/NfmNameRegistrationControllerFormProvider.scala:29]: are you sure we require 1000 char validation for name?

File: conf/application.conf (hunk 1 above)
Comment (javed-iqbal1) [reply conf/application.conf]: removed

File: conf/application.conf (hunk 2 above)
Comment (javed-iqbal1) [reply conf/application.conf]: removed

File: app/controllers/fm/IsNfmUKBasedController.scala [hunk 9]
```diff
@@ -64,15 +64,35 @@ class IsNfmUKBasedController @Inject() (
...
+                          FilingMember(NfmRegistrationConfirmation.No, Some(value), isNFMnStatus = RowStatus.InProgress)
+                        )
+                    )
+                _ <- userAnswersConnectors.save(updatedAnswers.id, Json.toJson(updatedAnswers.data))
+              } yield Redirect(routes.UnderConstructionController.onPageLoad)
+
+          }
       )
   }
 }
```
[the remark this answers is not in the document]
Comment (javed-iqbal1) [reply app/controllers/fm/IsNfmUKBasedController.scala:109]: added now

File: app/controllers/fm/NfmNameRegistrationController.scala [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+          } yield Redirect(controllers.fm.routes.NfmRegisteredAddressController.onPageLoad(mode))
+        }
+      )
+  }
+
+  private def isPreviousPageDefined(request: DataRequest[AnyContent]): Boolean =
+    request.userAnswers
+      .get(NominatedFilingMemberPage)
+      .fold(false) { data =>
+        data.isNfmRegisteredInUK.getOrElse() == NfmRegisteredInUkConfirmation.No
```
Comment (AKKUM) [inline app/controllers/fm/NfmNameRegistrationController.scala:98]: why getOrElse is required? please handle if data is not there.

File: test/controllers/testdata/Pillar2TestData.scala [hunk 11]
```diff
@@ -81,6 +122,14 @@ trait Pillar2TestData {
     postalCode = Some("VR11 3PA"),
     countryCode = "GB"
   )
+  val validNfmRegisteredAddressed = new NfmRegisteredAddress(
```
Comment (nitin-srivastav) [inline test/controllers/testdata/Pillar2TestData.scala:125]: spelling needs to be corrected

**Review by nitin-srivastav — DISMISSED** [review]

**Pull request merged**